### PR TITLE
Use the new multipart APIs

### PR DESCRIPTION
**Pull request by victorhurdugaci**
Depends on: https://github.com/aspnet/HttpAbstractions/pull/666

Please review @kichalla @Tratcher @rynowak

### CODE REVIEW

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 1]
```diff
@@ -66,9 +63,8 @@ public IActionResult FormToUpload()
                 ContentDispositionHeaderValue contentDisposition;
                 ContentDispositionHeaderValue.TryParse(section.ContentDisposition, out contentDisposition);
 
-                if (MultipartRequestHelper.HasFileContentDisposition(contentDisposition))
+                if (contentDisposition.IsFileContentDisposition())
                 {
-                    var name = HeaderUtilities.RemoveQuotes(contentDisposition.Name) ?? string.Empty;
```
Comment (kichalla) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: is this not required?

Comment (victorhurdugaci) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: Not unless reading the Name property has side effects :) `name` is not used in this `if` block

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 2]
```diff
@@ -47,7 +43,9 @@ public IActionResult FormToUpload()
         [ValidateAntiForgeryToken]
         public async Task<IActionResult> Upload()
         {
-            if (!MultipartRequestHelper.IsMultipartContentType(Request.ContentType))
+            var requestMediaType = MediaTypeHeaderValue.Parse(Request.ContentType);
```
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: is requestMediaType used anywhere else in this method? If not, why not have the extension directly on HttpResponse?

Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: nevermind..

**Comment by kichalla** [comment]
:shipit:

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 3]
```diff
@@ -66,9 +63,8 @@ public IActionResult FormToUpload()
                 ContentDispositionHeaderValue contentDisposition;
                 ContentDispositionHeaderValue.TryParse(section.ContentDisposition, out contentDisposition);
 
-                if (MultipartRequestHelper.HasFileContentDisposition(contentDisposition))
+                if (contentDisposition.IsFileContentDisposition())
                 {
-                    var name = HeaderUtilities.RemoveQuotes(contentDisposition.Name) ?? string.Empty;
                     var fileName = HeaderUtilities.RemoveQuotes(contentDisposition.FileName) ?? string.Empty;
```
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: you didn't find a way to clean this up? you talked about creating something like IFormFile from the multipart section.

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 4]
```diff
@@ -90,9 +86,9 @@ public IActionResult FormToUpload()
                     // Do not limit the key name length here because the mulipart headers length
                     // limit is already in effect.
                     var key = HeaderUtilities.RemoveQuotes(contentDisposition.Name);
-                    MediaTypeHeaderValue mediaType;
-                    MediaTypeHeaderValue.TryParse(section.ContentType, out mediaType);
-                    var encoding = FilterEncoding(mediaType?.Encoding);
+                    MediaTypeHeaderValue sectionMediaType;
```
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: Can this all get collapsed under a MultipartSection.ReadAsString() extension?

File: samples/Mvc.FileUpload/Controllers/FileController.cs (hunk 3 above)
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: This also needs to check FileNameStar

**Comment by victorhurdugaci** [comment]
Updated

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 5]
```diff
@@ -56,23 +54,17 @@ public IActionResult FormToUpload()
...
             var reader = new MultipartReader(boundary, HttpContext.Request.Body);
 
             var section = await reader.ReadNextSectionAsync();
             while (section != null)
             {
-                ContentDispositionHeaderValue contentDisposition;
-                ContentDispositionHeaderValue.TryParse(section.ContentDisposition, out contentDisposition);
-
-                if (MultipartRequestHelper.HasFileContentDisposition(contentDisposition))
+                var fileContentSection = FileContentMultipartSection.CreateFromMultipartSection(section);
```
Comment (kichalla) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: This improves the user experience. BTW can the multipart reader do the check for filecontent and instead return an instance of `FileContentMultipartSection`? (that would mean this section inheriting from MulitpartSection)

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 6]
```diff
@@ -56,23 +54,17 @@ public IActionResult FormToUpload()
...
+                if (fileContentSection != null)
                 {
-                    var name = HeaderUtilities.RemoveQuotes(contentDisposition.Name) ?? string.Empty;
-                    var fileName = HeaderUtilities.RemoveQuotes(contentDisposition.FileName) ?? string.Empty;
+                    var fileName = fileContentSection.FileName;
 
-                    // Here the uploaded file is being copied to local disk but you can also for example, copy the
-                    // stream directly to let's say Azure blob storage
                     targetFilePath = Path.Combine(_hostingEnvironment.ContentRootPath, Guid.NewGuid().ToString());
                     using (var targetStream = System.IO.File.Create(targetFilePath))
```
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: seciton.Body -> fileContentSection.FileStream

File: samples/Mvc.FileUpload/Controllers/FileController.cs (hunk 5 above)
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: Don't do that, the return type would still be MultipartSection and you'd have to dowcast to FileContentMultipartSection.

Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: I like FileContentMultipartSection, but not the magic return null if it doesn't match. How about
```
if (section.IsFile())
{
  var fileSection = section.ToFileSection();
  ...
}
```

**Comment by victorhurdugaci** [comment]
Updated

**Comment by Tratcher** [comment]
:shipit:

**Comment by victorhurdugaci** [comment]
Updated to the latest API

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 7]
```diff
@@ -56,58 +54,40 @@ public IActionResult FormToUpload()
...
             {
-                ContentDispositionHeaderValue contentDisposition;
-                ContentDispositionHeaderValue.TryParse(section.ContentDisposition, out contentDisposition);
-
-                if (MultipartRequestHelper.HasFileContentDisposition(contentDisposition))
-                {
-                    var name = HeaderUtilities.RemoveQuotes(contentDisposition.Name) ?? string.Empty;
-                    var fileName = HeaderUtilities.RemoveQuotes(contentDisposition.FileName) ?? string.Empty;
+                // This will reparse the content disposition header
+                // Create a
```
Comment (Tratcher) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: ?

**Comment by Tratcher** [comment]
:shipit:

File: Entropy.sln [hunk 8]
```diff
@@ -133,6 +133,18 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Mvc.TagHelperSample.Web", "
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Mvc.FileUpload", "samples\Mvc.FileUpload\Mvc.FileUpload.xproj", "{71EC0992-40E6-4643-9467-BB7A46218639}"
 EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.Http", "..\HttpAbstractions\src\Microsoft.AspNetCore.Http\Microsoft.AspNetCore.Http.xproj", "{BCF0F967-8753-4438-BD07-AADCA9CE509A}"
```
Comment (JunTaoLuo) [inline Entropy.sln]: tsk tsk

File: samples/Mvc.FileUpload/Controllers/FileController.cs [hunk 9]
```diff
@@ -47,7 +44,8 @@ public IActionResult FormToUpload()
         [ValidateAntiForgeryToken]
         public async Task<IActionResult> Upload()
         {
-            if (!MultipartRequestHelper.IsMultipartContentType(Request.ContentType))
+            var multipartBoundary = Request.GetMultipartBoundry();
```
Comment (JunTaoLuo) [inline samples/Mvc.FileUpload/Controllers/FileController.cs]: tsk tsk

**Pull request merged**